dev/integration_tests/flutter_gallery/test_driver/transitions_perf_e2e.dart

```diff
@@ -0,0 +1,119 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:ui';
+
+import 'package:flutter/cupertino.dart';
+import 'package:flutter/widgets.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+import 'package:e2e/e2e.dart';
+
+import 'package:flutter_gallery/gallery/app.dart' show GalleryApp;
+import 'package:flutter_gallery/gallery/demos.dart';
+import 'package:flutter_gallery/demo_lists.dart';
+
+import 'e2e_util.dart' show watchPerformance;
+
+const List<String> kSkippedDemos = <String>[];
+
+// All of the gallery demos, identified as "title@category".
+//
+// These names are reported by the test app, see _handleMessages()
+// in transitions_perf.dart.
+List<String> _allDemos = kAllGalleryDemos.map(
+  (GalleryDemo demo) => '${demo.title}@${demo.category.name}',
+).toList();
+
+/// Scrolls each demo menu item into view, launches it, then returns to the
+/// home screen twice.
+Future<void> runDemos(List<String> demos, WidgetTester tester) async {
+  final Finder demoList = find.byType(Scrollable);
+  String currentDemoCategory;
+
+  for (final String demo in demos) {
```

Comment:
A lot of the following code is very similar to `runDemo` in https://github.com/flutter/flutter/blob/master/dev/integration_tests/flutter_gallery/test_driver/transitions_perf_test.dart.

As we plan to preserve both Flutter driver timeline perf tests and e2e FrameTiming perf tests for a while (partially because timeline still has a lot more information that `FrameTiming` doesn't), it's probably better to use a single piece of code in both places. Maybe now is a good opportunity to let `FlutterDriver` implements `WidgetController` so we can share the code?

This would guarantee that driver tests and e2e tests are measuring similar animations, and avoid the risk that someone changes one test and forget to update another.

Reply:
It's hard. `WidgetController` has everything depending on a binding (it can be a test binding or a real binding), but `FlutterDriver` doesn't have a binding. And the finder in `WidgetController` and that in `FlutterDriver` also doesn't share base class.

Reply:
So to do that, the path in my opinion would be to write a mixin separating the controlling part (not the setting up part) of `WidgetController` and share this with `FlutterDriver`, and similarly for both finders. Even with that, there are some APIs doing the same thing but was named differently, meaning the result will be a breaking change. If we still consider abandon `flutter_driver` for driving the test, I don't think it will be worth it.

Reply:
Hmm... your concerns seems to be very legitimate. In that case, for Flutter driver timeline perf tests, can we just control the test inside `transitions_perf.dart` which is running on device instead of `transitions_perf_test.dart`? Then it seems that the host side `transitions_perf_test.dart` only needs to send a signal to device to start the control sequence (and the host will start capturing the timeline tracing events). The device-side code in `transitions_perf.dart` should be shareable with the E2E test?

Reply:
Haven't tried but sounds do-able. If that's the case the issue I mentioned above about using `WidgetController` and thus #62640 becomes necessary, and the measuring result may be different then the original ones because it's almost a new test (a host-driven and self-driven hybrid test). Will try tomorrow.

Reply:
https://github.com/flutter/flutter/pull/62064/commits/3e83f94d0c67a46bcb653966bd49cd208b6b2ea7 is an implementation of making `flutter drive` test share the driving code with the e2e test (making it a hybrid of host-driven and self-driven).